[python] Segfault with recursive untyped function

ESBMC 7.11.0 (commit hash: 58f7bb71d4) segfaults on this program:

```python
def f(x):
    if x == 0:
        return 0
    return f(x - 1)

assert f(3) == 0
```

Output: 
```bash
ESBMC version 7.11.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
Parsing main.py

Type checking warning:
main.py:2: error: Function is missing a type annotation  [no-untyped-def]
main.py:5: error: Call to untyped function "f" in typed context  [no-untyped-call]
main.py:7: error: Call to untyped function "f" in typed context  [no-untyped-call]
Found 3 errors in 1 file (checked 1 source file)

[1]    114666 segmentation fault (core dumped)  esbmc main.py
```

--- a/regression/python/github_3197/main.py
+++ b/regression/python/github_3197/main.py
@@ -1,0 +1,6 @@
+def f(x):
+    if x == 0:
+        return 0
+    return f(x - 1)
+
+assert f(3) == 0

--- a/regression/python/github_3197/test.desc
+++ b/regression/python/github_3197/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+
+^VERIFICATION SUCCESSFUL$

--- a/regression/python/infer-func-no-return/main.py
+++ b/regression/python/infer-func-no-return/main.py
@@ -10,23 +10,6 @@ def get_float():
 def get_boolean():
     return True
 
-# ===============================================
-# Functions with wrong return annotations
-# to test type inference
-# ===============================================
-
-def int_float() -> int: # Should infer to float
-  return 0.5
-
-
-def float_bool() -> float: # Should infer to bool
-  return False
-
-
-def bool_int() -> bool: # Should infer to int
-  return 3
-
-
 # ==============================
 # Test Cases with Assertions
 # ==============================
@@ -41,12 +24,3 @@ assert b == 0.5
 c = get_boolean()
 assert c == True
 
-# With incorrect annotations
-d = int_float()
-assert d == 0.5
-
-e = float_bool()
-assert e == False
-
-f = bool_int()
-assert f == 3

--- a/regression/python/infer-func-no-return2/main.py
+++ b/regression/python/infer-func-no-return2/main.py
@@ -1,0 +1,30 @@
+# ===============================================
+# Functions with wrong return annotations
+# to test type inference
+# ===============================================
+
+def int_float() -> int:
+  return 0.5
+
+
+def float_bool() -> float:
+  return False
+
+
+def bool_int() -> bool:
+  return 3
+
+
+# ==============================
+# Test Cases with Assertions
+# ==============================
+
+# With incorrect annotations
+d = int_float()
+assert d == 0.5
+
+e = float_bool()
+assert e == False
+
+f = bool_int()
+assert f == 3

--- a/regression/python/infer-func-no-return2/test.desc
+++ b/regression/python/infer-func-no-return2/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--override-return-annotation
+^VERIFICATION SUCCESSFUL$

--- a/regression/python/infer-func-no-return_fail/main.py
+++ b/regression/python/infer-func-no-return_fail/main.py
@@ -1,0 +1,30 @@
+# ===============================================
+# Functions with wrong return annotations
+# to test type inference
+# ===============================================
+
+def int_float() -> int:
+  return 0.5
+
+
+def float_bool() -> float:
+  return False
+
+
+def bool_int() -> bool:
+  return 3
+
+
+# ==============================
+# Test Cases with Assertions
+# ==============================
+
+# With incorrect annotations
+d = int_float()
+assert d == 0.5
+
+e = float_bool()
+assert e == False
+
+f = bool_int()
+assert f == 3

--- a/regression/python/infer-func-no-return_fail/test.desc
+++ b/regression/python/infer-func-no-return_fail/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+
+^VERIFICATION FAILED$

--- a/scripts/check_python_tests.sh
+++ b/scripts/check_python_tests.sh
@@ -44,6 +44,7 @@ ignored_dirs=(
   "github_3090_5"
   "github_3090_5_fail"
   "global"
+  "infer-func-no-return_fail"
   "integer_squareroot_fail"
   "int_from_bytes"
   "input"

--- a/src/python-frontend/python_annotation.h
+++ b/src/python-frontend/python_annotation.h
@@ -558,10 +558,13 @@ private:
     add_annotation(function_element);
 
     auto return_node = json_utils::find_return_node(function_element["body"]);
-    if (
-      !return_node.empty() &&
-      (function_element["returns"].is_null() ||
-       config.options.get_bool_option("override-return-annotation")))
+
+    // Check if we should override the return annotation
+    bool should_override =
+      config.options.get_bool_option("override-return-annotation");
+    bool has_no_annotation = function_element["returns"].is_null();
+
+    if (!return_node.empty() && (has_no_annotation || should_override))
     {
       std::string inferred_type;
       if (
@@ -816,10 +819,14 @@ private:
       {
         if (elem["_type"] == "FunctionDef" && elem["name"] == func_name)
         {
+          // When override flag is set, always infer instead of using annotation
+          bool should_override =
+            config.options.get_bool_option("override-return-annotation");
+
           // Check if function has explicit return type annotation
           if (
-            elem.contains("returns") && !elem["returns"].is_null() &&
-            elem["returns"].contains("id"))
+            !should_override && elem.contains("returns") &&
+            !elem["returns"].is_null() && elem["returns"].contains("id"))
           {
             functions_in_analysis_.erase(func_name);
             return elem["returns"]["id"];
@@ -857,59 +864,68 @@ private:
       }
     }
 
+    // Check override flag
+    bool should_override =
+      config.options.get_bool_option("override-return-annotation");
+
     // Search the top-level AST body for a matching function definition
     for (const Json &elem : ast["body"])
     {
       if (elem["_type"] == "FunctionDef" && elem["name"] == func_name)
       {
-        // Check if function has a return type annotation first (before inferring)
-        if (elem.contains("returns") && !elem["returns"].is_null())
+        // If override is set, skip annotation and go straight to inference
+        if (!should_override)
         {
-          const auto &returns = elem["returns"];
-
-          // Handle different return type annotation structures
-          if (returns.contains("_type"))
+          // Check if function has a return type annotation first (before inferring)
+          if (elem.contains("returns") && !elem["returns"].is_null())
           {
-            const std::string &return_type = returns["_type"];
+            const auto &returns = elem["returns"];
 
-            // Handle Subscript type (e.g., List[int], Dict[str, int])
-            if (return_type == "Subscript")
+            // Handle different return type annotation structures
+            if (returns.contains("_type"))
             {
-              functions_in_analysis_.erase(func_name);
-              if (returns.contains("value") && returns["value"].contains("id"))
-                return returns["value"]["id"];
-              else
-                return "Any"; // Default for complex subscript types
-            }
-            // Handle Constant type (e.g., None)
-            else if (return_type == "Constant")
-            {
-              functions_in_analysis_.erase(func_name);
-              if (returns.contains("value") && returns["value"].is_null())
-                return "NoneType";
-              else if (returns.contains("value"))
-                return "Any"; // Other constant types
-            }
-            // Handle other annotation types
-            else
-            {
-              // Try to extract id if it exists
-              if (returns.contains("id"))
+              const std::string &return_type = returns["_type"];
+
+              // Handle Subscript type (e.g., List[int], Dict[str, int])
+              if (return_type == "Subscript")
               {
                 functions_in_analysis_.erase(func_name);
-                return returns["id"];
+                if (
+                  returns.contains("value") && returns["value"].contains("id"))
+                  return returns["value"]["id"];
+                else
+                  return "Any"; // Default for complex subscript types
+              }
+              // Handle Constant type (e.g., None)
+              else if (return_type == "Constant")
+              {
+                functions_in_analysis_.erase(func_name);
+                if (returns.contains("value") && returns["value"].is_null())
+                  return "NoneType";
+                else if (returns.contains("value"))
+                  return "Any"; // Other constant types
+              }
+              // Handle other annotation types
+              else
+              {
+                // Try to extract id if it exists
+                if (returns.contains("id"))
+                {
+                  functions_in_analysis_.erase(func_name);
+                  return returns["id"];
+                }
               }
             }
-          }
-          // Handle case where returns exists but doesn't have expected structure
-          else
-          {
-            log_warning(
-              "Unrecognized return type annotation for function "
-              "{}",
-              func_name);
-            functions_in_analysis_.erase(func_name);
-            return "Any"; // Safe default
+            // Handle case where returns exists but doesn't have expected structure
+            else
+            {
+              log_warning(
+                "Unrecognized return type annotation for function "
+                "{}",
+                func_name);
+              functions_in_analysis_.erase(func_name);
+              return "Any"; // Safe default
+            }
           }
         }
 

--- a/src/python-frontend/python_annotation.h
+++ b/src/python-frontend/python_annotation.h
@@ -142,6 +142,85 @@ public:
   }
 
 private:
+  // Infer return type from non-recursive return statements
+  std::string
+  infer_from_return_statements(const Json &body, const std::string &func_name)
+  {
+    for (const Json &stmt : body)
+    {
+      // Found a return statement
+      if (stmt["_type"] == "Return" && !stmt["value"].is_null())
+      {
+        const Json &return_val = stmt["value"];
+
+        // Skip recursive calls
+        if (
+          return_val["_type"] == "Call" && return_val.contains("func") &&
+          return_val["func"]["_type"] == "Name" &&
+          return_val["func"]["id"] == func_name)
+        {
+          continue; // Skip this recursive call
+        }
+
+        // Try to infer type from the return value
+        if (return_val["_type"] == "Constant")
+        {
+          const auto &value = return_val["value"];
+          if (value.is_number_integer())
+            return "int";
+          if (value.is_number_float())
+            return "float";
+          if (value.is_boolean())
+            return "bool";
+          if (value.is_string())
+            return "str";
+          if (value.is_null())
+            return "NoneType";
+        }
+        else if (return_val["_type"] == "Name")
+        {
+          // Try to find the variable's type
+          Json var_node =
+            json_utils::find_var_decl(return_val["id"], func_name, ast_);
+          if (
+            !var_node.empty() && var_node.contains("annotation") &&
+            !var_node["annotation"].is_null() &&
+            var_node["annotation"].contains("id"))
+          {
+            return var_node["annotation"]["id"];
+          }
+        }
+        else if (return_val["_type"] == "BinOp")
+        {
+          // For binary operations, try to infer from operands
+          Json dummy_stmt = {{"value", return_val}};
+          std::string type = get_type_from_binary_expr(dummy_stmt, ast_);
+          if (!type.empty())
+            return type;
+        }
+      }
+
+      // Recursively check nested blocks (if/while/for/try)
+      if (stmt.contains("body") && stmt["body"].is_array())
+      {
+        std::string nested_type =
+          infer_from_return_statements(stmt["body"], func_name);
+        if (!nested_type.empty())
+          return nested_type;
+      }
+
+      if (stmt.contains("orelse") && stmt["orelse"].is_array())
+      {
+        std::string nested_type =
+          infer_from_return_statements(stmt["orelse"], func_name);
+        if (!nested_type.empty())
+          return nested_type;
+      }
+    }
+
+    return ""; // Couldn't infer
+  }
+
   // Method to infer and annotate unannotated function parameters
   void infer_parameter_types(Json &function_element)
   {
@@ -761,13 +840,51 @@ private:
   std::string
   get_function_return_type(const std::string &func_name, const Json &ast)
   {
+    // Guard against infinite recursion
+    if (functions_in_analysis_.count(func_name) > 0)
+    {
+      // Function is calling itself: try to infer from non-recursive return statements
+      for (const Json &elem : ast["body"])
+      {
+        if (elem["_type"] == "FunctionDef" && elem["name"] == func_name)
+        {
+          // Check if function has explicit return type annotation
+          if (
+            elem.contains("returns") && !elem["returns"].is_null() &&
+            elem["returns"].contains("id"))
+          {
+            functions_in_analysis_.erase(func_name);
+            return elem["returns"]["id"];
+          }
+
+          // Try to infer from return statements (excluding recursive calls)
+          std::string inferred =
+            infer_from_return_statements(elem["body"], func_name);
+          if (!inferred.empty())
+          {
+            functions_in_analysis_.erase(func_name);
+            return inferred;
+          }
+
+          // No annotation and can't infer: return empty to avoid crash
+          functions_in_analysis_.erase(func_name);
+          return "";
+        }
+      }
+      functions_in_analysis_.erase(func_name);
+      return "";
+    }
+
+    // Add function to set before analysis
+    functions_in_analysis_.insert(func_name);
+
     // Get type from nondet_<type> functions
     if (func_name.find("nondet_") == 0)
     {
       size_t last_underscore_pos = func_name.find_last_of('_');
       if (last_underscore_pos != std::string::npos)
       {
-        // Return the substring from the position after the last underscore to the end
+        functions_in_analysis_.erase(func_name);
         return func_name.substr(last_underscore_pos + 1);
       }
     }
@@ -777,16 +894,7 @@ private:
     {
       if (elem["_type"] == "FunctionDef" && elem["name"] == func_name)
       {
-        // Try to infer return type from actual return statements
-        auto return_node = json_utils::find_return_node(elem["body"]);
-        if (!return_node.empty())
-        {
-          std::string inferred_type;
-          infer_type(return_node, elem, inferred_type);
-          return inferred_type;
-        }
-
-        // Check if function has a return type annotation
+        // Check if function has a return type annotation first (before inferring)
         if (elem.contains("returns") && !elem["returns"].is_null())
         {
           const auto &returns = elem["returns"];
@@ -799,6 +907,7 @@ private:
             // Handle Subscript type (e.g., List[int], Dict[str, int])
             if (return_type == "Subscript")
             {
+              functions_in_analysis_.erase(func_name);
               if (returns.contains("value") && returns["value"].contains("id"))
                 return returns["value"]["id"];
               else
@@ -807,6 +916,7 @@ private:
             // Handle Constant type (e.g., None)
             else if (return_type == "Constant")
             {
+              functions_in_analysis_.erase(func_name);
               if (returns.contains("value") && returns["value"].is_null())
                 return "NoneType";
               else if (returns.contains("value"))
@@ -817,7 +927,10 @@ private:
             {
               // Try to extract id if it exists
               if (returns.contains("id"))
+              {
+                functions_in_analysis_.erase(func_name);
                 return returns["id"];
+              }
             }
           }
           // Handle case where returns exists but doesn't have expected structure
@@ -827,10 +940,23 @@ private:
               "Unrecognized return type annotation for function "
               "{}",
               func_name);
+            functions_in_analysis_.erase(func_name);
             return "Any"; // Safe default
           }
         }
+
+        // Try to infer return type from actual return statements
+        auto return_node = json_utils::find_return_node(elem["body"]);
+        if (!return_node.empty())
+        {
+          std::string inferred_type;
+          infer_type(return_node, elem, inferred_type);
+          functions_in_analysis_.erase(func_name);
+          return inferred_type;
+        }
+
         // If function has no explicit return statements, assume void/None
+        functions_in_analysis_.erase(func_name);
         return "NoneType";
       }
     }
@@ -843,7 +969,10 @@ private:
       lambda_elem = find_lambda_in_body(func_name, (*current_func)["body"]);
 
     if (!lambda_elem.empty())
+    {
+      functions_in_analysis_.erase(func_name);
       return infer_lambda_return_type(lambda_elem);
+    }
 
     // Get type from imported functions
     try
@@ -866,14 +995,17 @@ private:
           {
             // It might be a class constructor (__init__ returns None)
             // Return the class name as the type
+            functions_in_analysis_.erase(func_name);
             return func_name;
           }
 
+          functions_in_analysis_.erase(func_name);
           return func_info.return_type_;
         }
         catch (std::runtime_error &)
         {
           // If get_function fails, it might be a class
+          functions_in_analysis_.erase(func_name);
           return func_name;
         }
       }
@@ -886,8 +1018,11 @@ private:
     auto it = builtin_functions.find(func_name);
     if (it != builtin_functions.end())
     {
+      functions_in_analysis_.erase(func_name);
       return it->second;
     }
+
+    functions_in_analysis_.erase(func_name);
 
     std::ostringstream oss;
     oss << "Function \"" << func_name << "\" not found (" << python_filename_
@@ -2014,4 +2149,5 @@ private:
   std::string python_filename_;
   bool filter_global_elements_ = false;
   std::vector<Json> referenced_global_elements;
+  std::set<std::string> functions_in_analysis_;
 };

--- a/src/python-frontend/python_annotation.h
+++ b/src/python-frontend/python_annotation.h
@@ -162,42 +162,10 @@ private:
           continue; // Skip this recursive call
         }
 
-        // Try to infer type from the return value
-        if (return_val["_type"] == "Constant")
-        {
-          const auto &value = return_val["value"];
-          if (value.is_number_integer())
-            return "int";
-          if (value.is_number_float())
-            return "float";
-          if (value.is_boolean())
-            return "bool";
-          if (value.is_string())
-            return "str";
-          if (value.is_null())
-            return "NoneType";
-        }
-        else if (return_val["_type"] == "Name")
-        {
-          // Try to find the variable's type
-          Json var_node =
-            json_utils::find_var_decl(return_val["id"], func_name, ast_);
-          if (
-            !var_node.empty() && var_node.contains("annotation") &&
-            !var_node["annotation"].is_null() &&
-            var_node["annotation"].contains("id"))
-          {
-            return var_node["annotation"]["id"];
-          }
-        }
-        else if (return_val["_type"] == "BinOp")
-        {
-          // For binary operations, try to infer from operands
-          Json dummy_stmt = {{"value", return_val}};
-          std::string type = get_type_from_binary_expr(dummy_stmt, ast_);
-          if (!type.empty())
-            return type;
-        }
+        // Reuse get_argument_type to infer the return value type
+        std::string inferred_type = get_argument_type(return_val);
+        if (!inferred_type.empty())
+          return inferred_type;
       }
 
       // Recursively check nested blocks (if/while/for/try)

